modules/database/src/std/rec/stdstringinRecord.cpp

```diff
@@ -0,0 +1,240 @@
+
+#ifndef USE_TYPED_RSET
+#  define USE_TYPED_RSET
+#endif
+#ifndef USE_TYPED_DSET
+#  define USE_TYPED_DSET
+#endif
+
+#include <dbAccess.h>
+#include <recSup.h>
+#include <recGbl.h>
+#include <dbEvent.h>
+
+// include by stdstringinRecord.h
+#include "epicsTypes.h"
+#include "link.h"
+#include "epicsMutex.h"
+#include "ellLib.h"
+#include "devSup.h"
+#include "epicsTime.h"
+
+#define epicsExportSharedSymbols
+
+#define GEN_SIZE_OFFSET
+#include <stdstringinRecord.h>
+#undef  GEN_SIZE_OFFSET
+#include <epicsExport.h>
+
+extern "C" {
+VFieldType vfStdString = {"std::string"};
+}
+
+namespace  {
+
+ELLLIST vfList = ELLLIST_INIT;
+VFieldTypeNode vfStrNode;
+
+long initialize()
+{
+    vfStrNode.vtype = &vfStdString;
+    ellAdd(&vfList, &vfStrNode.node);
+    return 0;
+}
+
+long init_record(struct dbCommon *pcommon, int pass)
+{
+    stdstringinRecord *prec = (stdstringinRecord*)pcommon;
+    stdstringindset *pdset = (stdstringindset *)(prec->dset);
+
+    if (!pdset) {
+        recGblRecordError(S_dev_noDSET, prec, "stdstringin: init_record");
+        return S_dev_noDSET;
+    }
+
+    if(pass==0) {
+        new (&prec->val) std::string();
+        new (&prec->oval) std::string();
+
+        if(pdset->common.init_record) {
+            long ret = pdset->common.init_record(pcommon);
+            if(ret)
+                return ret;
+
+            prec->oval = prec->val;
+        }
+    }
+    return 0;
+}
+
+long readValue(stdstringinRecord *prec,
+               stdstringindset *pdset)
+{
+    return pdset->read_stdstringin(prec);
+}
+
+void monitor(stdstringinRecord *prec)
+{
+    int monitor_mask = recGblResetAlarms(prec);
+
+    if (prec->oval != prec->val) {
+        monitor_mask |= DBE_VALUE | DBE_LOG;
+        prec->oval = prec->val;
+    }
+
+    if (monitor_mask) {
+        db_post_events(prec, &prec->val, monitor_mask);
+        db_post_events(prec, &prec->oval, monitor_mask);
+    }
+}
+
+long process(struct dbCommon *pcommon)
+{
+    stdstringinRecord *prec = (stdstringinRecord*)pcommon;
+    stdstringindset *pdset = (stdstringindset *)(prec->dset);
+    unsigned char    pact=prec->pact;
+    long status;
+
+    if( (pdset==NULL) || (pdset->read_stdstringin==NULL) ) {
+        prec->pact=TRUE;
+        recGblRecordError(S_dev_missingSup,(void *)prec,"read_stdstringin");
+        return(S_dev_missingSup);
+    }
+
+    status=readValue(prec, pdset); /* read the new value */
+    /* check if device support set pact */
+    if ( !pact && prec->pact ) return(0);
+
+    prec->pact = TRUE;
+    recGblGetTimeStamp(prec);
+
+    /* check event list */
+    monitor(prec);
+    /* process the forward scan link record */
+    recGblFwdLink(prec);
+
+    prec->pact=FALSE;
+    return status;
+}
+
+long cvt_dbaddr(DBADDR *paddr)
+{
+    // for both VAL and OVAL
+
+    // we don't allow dbPut() via DBF_CHAR.
+    // std::string::c_str() warns of dire consequences following modifications
+    paddr->ro = 1;
```

Comment:
I think you can have dbAccess prevent puts by setting `paddr->special = SPC_NOMOD` here, but I don't know what that would do with your `put_vfield()` additions.

Reply:
Would this actually prevent `dbPut()` from working?  From my reading of the code, I don't think so.  At least no without  Also, I still want to distinguish fields which `dbPut()` should never be allowed (even through `put_vfield()`).